public/static/settings.js

```diff
@@ -0,0 +1,83 @@
+// This function handles the toggling of selections of all upstream search engines 
+// options in the settings page under the tab engines.
+function toggleAllSelection() {
+  document
+    .querySelectorAll('.engine')
+    .forEach(
+      (engine_checkbox) =>
+        (engine_checkbox.checked =
+          document.querySelector('.select_all').checked)
+    )
+}
+
+// This function adds the functionality to sidebar buttons to only show settings 
+// related to that tab.
+function setActiveTab(current_tab) {
+  document
+    .querySelectorAll('.tab')
+    .forEach((tab) => tab.classList.remove('active'))
+  document
+    .querySelectorAll('.btn')
+    .forEach((tab) => tab.classList.remove('active'))
+  current_tab.classList.add('active')
+  document
+    .querySelector(`.${current_tab.innerText.toLowerCase().replace(' ', '_')}`)
+    .classList.add('active')
+}
+
+// This function adds the functionality to save all the user selected preferences
+// to be saved in a cookie on the users machine.
+function setClientSettings() {
+  let cookie_dictionary = new Object()
+  document.querySelectorAll('select').forEach((select_tag) => {
+    if (select_tag.name === 'themes') {
+      cookie_dictionary['theme'] = select_tag.value
+    } else if (select_tag.name === 'colorschemes') {
+      cookie_dictionary['colorscheme'] = select_tag.value
+    }
+  })
+  let engines = []
+  document.querySelectorAll('.engine').forEach((engine_checkbox) => {
+    if (engine_checkbox.checked === true) {
+      engines.push(engine_checkbox.parentNode.parentNode.innerText.trim())
+    }
+  })
+  cookie_dictionary['engines'] = engines
+  let expiration_date = new Date()
+  expiration_date.setFullYear(expiration_date.getFullYear() + 1)
+  document.cookie = `appCookie=${JSON.stringify(
+    cookie_dictionary
+  )}; expires=${expiration_date.toUTCString()}`
+
+  document.querySelector('.message').innerText =
+    '✅ The settings have been saved sucessfully!!'
+
+  setTimeout(() => {
+    document.querySelector('.message').innerText = ''
+  }, 10000)
+}
```

Comment:
The third function, setClientSettings, uses querySelectorAll which can be slow if there are many elements to search through. It also uses forEach which can be slow if there are many elements to iterate through. A more performant solution would be to use a for loop instead of forEach.
```suggestion
function setClientSettings() {
  const cookieDictionary = {};
  const selectTags = document.querySelectorAll('select');
  for (let i = 0; i < selectTags.length; i++) {
    const selectTag = selectTags[i];
    if (selectTag.name === 'themes') {
      cookieDictionary['theme'] = selectTag.value;
    } else if (selectTag.name === 'colorschemes') {
      cookieDictionary['colorscheme'] = selectTag.value;
    }
  }
  const engines = [];
  const engineCheckboxes = document.querySelectorAll('.engine');
  for (let i = 0; i < engineCheckboxes.length; i++) {
    const engineCheckbox = engineCheckboxes[i];
    if (engineCheckbox.checked === true) {
      engines.push(engineCheckbox.parentNode.parentNode.innerText.trim());
    }
  }
  cookieDictionary['engines'] = engines;
  const expirationDate = new Date();
  expirationDate.setFullYear(expirationDate.getFullYear() + 1);
  document.cookie = `appCookie=${JSON.stringify(cookieDictionary)}; expires=${expirationDate.toUTCString()}`;
  document.querySelector('.message').innerText = '✅ The settings have been saved sucessfully!!';
  setTimeout(() => {
    document.querySelector('.message').innerText = '';
  }, 10000);
}
```